modules/caddyhttp/app.go

```diff
@@ -486,7 +460,24 @@ func (app *App) Start() error {
 					}
 				}
 			}
-		} else {
+		}
+
+		// configure the http versions the server will serve
+		if srv.protocol("h1") {
+			srv.server.Protocols.SetHTTP1(true)
+		}
+
+		if srv.protocol("h2") || srv.protocol("h2c") {
+			// skip setting h2 because if NextProtos is present, it's list of alpn versions will take precedence.
+			// it will always be present because http2.ConfigureServer will populate that field
+			// enabling h2c because some listener wrapper will wrap the connection that is no longer *tls.Conn
+			// However, we need to handle the case that if the connection is h2c but h2c is not enabled. We identify
+			// this type of connection by checking if it's behind a TLS listener wrapper or if it implements tls.ConnectionState.
+			srv.server.Protocols.SetUnencryptedHTTP2(true)
```

Comment:
Shouldn't this only be set if `h2c`, but not `h2`?

Reply:
This is deliberate. if the handshake succeeds with `h2`, it will be handled accordingly. `h2c` is enabled to handle listener wrappers.

Reply:
But if the user only configured `h2` and _not_ `h2c`, it'll still have `SetUnencryptedHTTP2(true)`, right? That doesn't seem correct.

Reply:
Or wait, you mean if there's LWs then it could appear as though the conn is no longer encrypted cause it's already been decrypted? Is that what you mean? Tricky.